README.md
<p align="center"> 
  <img src="img/icon_trn.svg" alt="TRN Logo" width="80px" height="80px">
</p>
<h1 align="center"> TrayNim </h1>
<h3 align="center"> A simple ray tracer written in Nim </h3>  

</br>

<p align="center"> 
  <img src="img/traynim_banner.gif" alt="Sample signal" width="80%" height="80%">
</p>

<!-- Add buttons here -->

![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/ottyanna/traynim?include_prereleases)
![GitHub last commit](https://img.shields.io/github/last-commit/ottyanna/traynim)
![GitHub issues](https://img.shields.io/github/issues-raw/ottyanna/traynim)
![GitHub pull requests](https://img.shields.io/github/issues-pr/ottyanna/traynim)
![GitHub](https://img.shields.io/github/license/ottyanna/traynim)
![GitHub Workflow Status](https://img.shields.io/github/workflow/status/ottyanna/traynim/BuildAndTest)
![GitHub status](https://img.shields.io/badge/status-WIP-informational)

Traynim is a CLI program written in Nim that lets you create photorealistic images using spheres and planes. 
It can also be used as a tool for converting PFM files to PNG, PPM, BMP and QOI.

To get a more detailed description of the source code, refer to the [documentation](https://ottyanna.github.io/traynim/htmldocs/theindex.html).

## Demo-Preview

Traynim can produce images like these with four different renderers: path tracer, on/off, point light and flat renderer respectively. [Here](https://ottyanna.github.io/traynim/gallery/gallery.html) you can find some more images.

<p align="center"> 
  <img src="img/demo/Antialiasing/pathTracingr4s16.png" alt="Sample signal" width="45%" height="50%">
  <img src="img/demo/spheres-perspective.gif" alt="Sample signal" width="45%" height="50%">
</p>

<p align="center">
  <img src="img/demo/Antialiasing/PointLightr4s16.png" alt="Sample signal" width="45%" height="50%">
  <img src="img/blurrystack.png" alt="Sample signal" width="45%" height="50%">
</p>

<p align="center">
  <img src="img/sw.gif" alt="Sample signal" width="45%" height="50%">
</p>



# Table of contents

- [Demo-Preview](#demo-preview)
- [Table of contents](#table-of-contents)
- [Installation](#installation)
- [Usage](#usage)
- [Examples](#examples)
- [Contribute](#contribute)
- [Release History](#release-history)
- [License](#license)
- [Footnotes](#footnotes)

## Installation
[(Back to top)](#table-of-contents)

### Dependencies
- [nim](https://github.com/nim-lang) >= 1.6.4
- [nimble](https://github.com/nim-lang/nimble)
- [ffmpeg](https://www.ffmpeg.org/) and [parallel](https://en.wikipedia.org/wiki/GNU_parallel) for the bash scripts



### Download and building
You can download the latest stable release [here](https://github.com/ottyanna/traynim/releases), and unpack it
   ``` sh
   $ tar -xvf /path/to/tar #or zip file -C /path/to/your/directory
   ```
or if you want, you can clone this repository
   ``` sh
   $ git clone https://github.com/ottyanna/traynim.git
   ```

### Testing
You can test if the code works fine by running the following command:
``` sh
   $ nimble test
   ```

## Usage
[(Back to top)](#table-of-contents)

```sh
$ nimble run traynim
```
will generate the executable.

To display the command-line help, you can use

```sh
$ ./traynim --help
                       
Usage:
  traynim {SUBCMD}  [sub-command options & parameters]
where {SUBCMD} is one of:
  help        print comprehensive or per-cmd help
  renderer    
  stack       
  pfm2format  
  demo        

traynim {-h|--help} or with no args at all prints this message.
traynim --help-syntax gives general cligen syntax help.
Run "traynim {help SUBCMD|SUBCMD --help}" to see help for just SUBCMD.
Run "traynim help" to get *comprehensive* help.
```

### `renderer`

To generate one image with default parameters (the default and example input file is [here](examples/example.txt)) use

```sh
$ ./traynim renderer
```

You can change the size of the image and the `png` output conversion luminosity by running `--help`.

As shown in [Demo-Preview](#demo-preview), it is possible to use different rendering methods by using the `-a` or `--algorithm` flag.

 Renderer |  Feature | Option
:-------------------------:|:-------------------------: | :------:
Simple Path Tracer  | Most photorealistic renderer (Uses Russian roulette method) | `-a=pathtracing` (Default)
Point Light Renderer  | The source light is considered a Dirac delta | `-a=pointlight`
Flat Renderer | The scene is traced without considering the light source, so it is in solid colors | `-a=flat`
On/off renderer | There are just two colors, one for the objects, one for the background | `-a=on/off`

While the path tracer is really slow, the flat, point light and on/off renderers are fast, so they are useful for debugging or to get an idea if the object positions are right. The time scale depends really on the scene but for a simple scene it can take up to 5 minutes for the path tracing while about 5 to 20 seconds for the others. 
<!--- Moreover, bear in mind that if you use a sphere for the sky the on/off renderer--> 

ANIMATIONS

You can generate the animations by tweaking [this](scripts/generateImage.sh) bash script, then just run

```sh
$ ./animation.sh NUM_OF_CORES >/dev/null
```

where `NUM_OF_CORES` are the number of cores to parallelize the generation.

### `stack`

Generating an image can take a long time, especially if you need to cancel out noise. You can reduce the time by using the [antialiasing](#antialiasing-in-case-of-flat-rendering) feature or you can just use the `stack` command that takes advantage of the seekability property of the chosen number generator (by setting a different `initSeq`). Once you render the same scene (`numOfFiles` times) with different noise (again by setting `-initSeq` to different values), the `stack` command merges them into the same image. 
The only thing you have to bear in mind is that the images must be named img0, img1, img2 etc (`inPfmFileName` is img in this case). This [script](scripts/stackMain.sh) makes this procedure automatic using [parallel](https://en.wikipedia.org/wiki/GNU_parallel).

You can also use this feature to reduce noise AND to make blurry images. [Here](#example-for-the-blurry-effect) you can see an example.

### `pfm2format`

To use the `pfm2format` feature, you have to bear in mind that just PNG, PPM, BMP and [QOI](https://en.wikipedia.org/wiki/QOI_(image_format)) formats are supported.


### `demo` (Deprecated)

The `demo` command is still usable, but since version 1.0.0, it's better to use files for scenes. Despite this, if you wish to create your images with `for` cycles, like the scene in the on/off example in the [demo_preview](#demo-preview) it's better to tweak directly the `demo` part of the code.

## Examples

### Antialiasing in case of flat rendering:

You can reduce the noise of the image by increasing the number of rays or by increasing the samples per pixel. The latter can be used with all the tracing algorithms.

The scene has a white light in coordinates (-50, 30, 30). The number of rays is always 4 (`-r=4`) and the luminosity is 0.8.

 `-s=1` |  `-s=4` | `-s=25`
:-------------------------:|:-------------------------:|:-------------------------: 
![](img/demo/Antialiasing/flatr4.png)  |  ![](img/demo/Antialiasing/flatr4s4.png) | ![](img/demo/Antialiasing/flatr4s25.png)
4 seconds | 16 seconds | 104 seconds

As you can see the noise is sensibly reduced by increasing the samplesPerPixel.

### Example of the noise reduction with `stack`

<p align="center"> 
  <img src="img/test0.png" alt="Sample signal" width="45%" height="50%">
  <img src="img/testDef.png" alt="Sample signal" width="45%" height="50%">
</p>

On the left there is one of the six images generated with different `initSeq` value.
On the right there is the "stacked" image: you can see how it gets less noisy.

### Example for the blurry effect

<p align="left"> 
  <img src="img/blurrystack.png" alt="blurrystack" width="45%" height="50%">
</p>

This image was obtained by rotating both spheres a little. In case of path tracing you could also use a different `initSeq` value, in order to reduce the overall noise. Refer to this [script](scripts/stack.sh) for an example.

### `--gamma` and `--factor` variation in `pfm2format` feature:

<!---maybe we need to write the site of the image--->

|    options     | `-f=0.15`             |  `-f=0.30` | `-f=0.50`
:---------:|:-------------------------:|:-------------------------:|:-------------------------: 
`-g=1.0` | ![](img/pfm2formatExamples/sample1.00.15.png)  |  ![](img/pfm2formatExamples/sample1.00.30.png) | ![](img/pfm2formatExamples/sample1.00.50.png) 
`-g=2.2` | ![](img/pfm2formatExamples/sample2.20.15.png) | ![](img/pfm2formatExamples/sample2.20.30.png)  |  ![](img/pfm2formatExamples/sample2.20.50.png)

You can compare different values for `gamma` and `factor` tweaking the bash script in [img/pfm2formatExamples](img/pfm2formatExamples/) folder<sup id="a1">[1](#f1)</sup>

## Contribute
[(Back to top)](#table-of-contents)

If you wish to contribute or you found any bug, feel free to open an issue or a pull request on the GitHub repository.

Or if you just want to show your images made with the program, contact us, we'll be happy to show them on the project website. 

## Release History
[(Back to top)](#table-of-contents)

See the [CHANGELOG.md](https://github.com/ottyanna/traynim/blob/master/CHANGELOG.md) file.

## License
[(Back to top)](#table-of-contents)

[GNU GPLv3](https://choosealicense.com/licenses/gpl-3.0/)

## Footnotes
<b id="f1">1</b> The sample image was taken from [here](https://filesamples.com/formats/pfm). [↩](#a1) 

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [1.0.1] - 22/06/2022

### Fixed
- Fixed an issue with missing pfm image [#10](https://github.com/ottyanna/traynim/pull/10)
- Fixed bug with stackMain.sh script

## [1.0.0] - 18/06/2022

## Added
- **Major update**: new `renderer` command to add scene from file 
- **Major update**: new `stack` command to reduce noise
- Added compiler to parse scene from file

## Deprecated
- `demo` command


## [0.3.0] - 25/05/2022

## Added
- **Major update**: Path Tracing renderer
- **Major update**: Flat renderer
- **Major update**: Point Light renderer
- Antialiasing feature
- Some pigments and BRDFs


## [0.2.0] - 11/05/2022

### Added
- **Major update**: new `demo` command
- implemented spheres and planes
- on/off rendering feature

### Changed 
- **Major update**: nicer CLI using [cligen](https://github.com/c-blake/cligen)

### Removed
- Removed old CLI

## [0.1.1] - 27/04/2022

### Fixed
-   Fixed an issue with the vertical order of the images [#4](https://github.com/ottyanna/traynim/pull/4)

## [0.1.0] - 30/03/2022
First release of the code

### Added
- Reading from PFM files
- Tone mapping
- Gamma correction
- Saving files in PNG, PPM, BMP and QOI formats

[0.1.0]: https://github.com/ottyanna/traynim/releases/tag/v0.1.0
[0.1.1]: https://github.com/ottyanna/traynim/releases/tag/v0.1.1
[0.2.0]: https://github.com/ottyanna/traynim/releases/tag/v0.2.0
[0.3.0]: https://github.com/ottyanna/traynim/releases/tag/v0.3.0
[1.0.0]: https://github.com/ottyanna/traynim/releases/tag/v1.0.0
[1.0.1]: https://github.com/ottyanna/traynim/releases/tag/v1.0.1

src/traynim.nim
#encoding: utf-8

#traynim is a ray tracer program written in Nim
#Copyright (C) 2022 Jacopo Fera, Anna Spanò

#This program is free software: you can redistribute it and/or modify
#it under the terms of the GNU General Public License as published by
#the Free Software Foundation, either version 3 of the License, or
#(at your option) any later version.

#This program is distributed in the hope that it will be useful,
#but WITHOUT ANY WARRANTY; without even the implied warranty of
#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#GNU General Public License for more details.

#You should have received a copy of the GNU General Public License
#along with this program.  If not, see <https://www.gnu.org/licenses/>.


import strutils, streams, cligen, sugar
from math import sqrt, pow
import tables
when compileOption("profiler"):
  import nimprof

import
    cameras,
    colors,
    geometry,
    hdrimages,
    imageTracer,
    lights,
    materials,
    std/monotimes,
    pcg,
    render,
    shapes,
    transformations,
    world,
    sceneFiles,
    options

proc buildVariablesTable(declareFloat:string):Table[string,float] =

    for item in splitWhitespace(declareFloat):
        
        var stringVariable = item.split(':')
        
        if stringVariable.len != 2:
            quit("Error: The format for variable declaration NAME:VALUE is not respected")
        else:
            var floatVar: float
            try:
                floatVar = stringVariable[1].parseFloat()
            except ValueError:
                quit("Error: The second term is not a floating-point")

            result[stringVariable[0]]=floatVar

# --------------RENDER--------------

proc renderer( inSceneName : string = "examples/example.txt",
               declareFloat:string ="",
               width = 640, height = 480,
               algorithm = "pathtracing",
               fileName = "demo", format = "png", 
               luminosity: float = 0.0, 
               samplesPerPixel = 1.0, 
               raysNum = 10, maxDepth = 3, 
               initState = 42, initSeq = 54)=



    let samplesPerSide = sqrt(samplesPerPixel).int
    if pow(samplesPerSide.float, 2.0) != samplesPerPixel:
        quit("Error: the number of samples per pixel ({samplePerPixel}) must be a perfect square")

    var inScene: Stream
    var variables:Table[string,float]
    try:
        if declareFloat=="":
            inScene = openFileStream(inSceneName,fmRead)
        else:
            variables = buildVariablesTable(declareFloat)
            inScene = openFileStream(inSceneName,fmRead)
    except IOError:
        quit("ERROR: " & getCurrentExceptionMsg())

    var inputStream = newInputStream(stream=inScene, fileName=inSceneName)
 
    var scene : Scene

    try:
        scene = parseScene(inputStream, variables)
    except GrammarError:
        quit("ERROR: " & getCurrentExceptionMsg() & " [GrammarError]")

    inScene.close()

    var image = newHDRImage(width, height)
    echo("Generating the ", width, "x", height,
            " image...")

    if scene.camera.isNone:
        quit("ERROR: camera not defined in input scene [GrammarError]")

    var tracer = newImageTracer(image, scene.camera.get, samplesPerSide)

    var renderer: Renderer

    case algorithm:
        of "on/off":
            echo("Using on/off renderer")
            renderer = newOnOffRenderer(scene.world, black, white)
        of "flat":
            echo("Using flat renderer")
            renderer = newFlatRenderer(scene.world)
        of "pathtracing":
            echo("Using pathtracing")
            var pcg = newPCG(initState = initState.uint64,
                    initSeq = initSeq.uint64)
            renderer = newPathTracer(
                world = scene.world,
                pcg = pcg,
                raysNum = raysNum,
                maxDepth = maxDepth
            )
        of "pointlight":
            echo("Using point-light tracer")
            renderer = newPointLightRenderer(world = scene.world,
                    backgroundColor = black)
        else:
            quit("Unknown renderer")

    let time = getMonoTime()
    tracer.fireAllRays(ray => call(renderer, ray))
    echo "Time taken: ", getMonoTime() - time

    # Save the HDR image
    let outPfm = newFileStream(fileName & ".pfm", fmWrite)
    tracer.image.writePfmImage(outPfm)
    echo "HDR demo image written to " & fileName & ".pfm"
    outPfm.close()

    # Apply tone-mapping to the image
    if luminosity == 0.0:
        tracer.image.normalizeImage(factor = 1.0) #use average luminosity
    else:
        tracer.image.normalizeImage(factor = 1.0, luminosity)

    tracer.image.clampImage()

    # Save the LDR image
    tracer.image.writeLdrImage(fileName & "." & format)
    echo "PNG demo image written to " & fileName & ".png"

# --------------STACK--------------

proc stack(inPfmFileName: string, numOfFiles : int,
        outputFileName: string, factor=0.2, gamma=1.0) =

    var imgs : seq[HDRimage] = @[]

    for i in 0..<numOfFiles:
        var inPfm :Stream 
        try:
            inPfm = newFileStream(inPfmFileName & $i & ".pfm" , fmRead)
        except IOError:
            quit("ERROR: " & getCurrentExceptionMsg())
        imgs.add(readPfmImage(inPfm))
        inPfm.close()

    var oImg = newHDRImage(imgs[0].width, imgs[0].height)

    var cumColor: Color = black

    for x in 0..<oImg.width:
        for y in 0..<oImg.height:
            cumColor=black
            for i in 0..<imgs.len:
                cumColor = cumColor + imgs[i].getPixel(x,y)
            oImg.setPixel(x,y,(cumColor/imgs.len.float))

    # Save the HDR image
    let outPfm = newFileStream(outputFileName & ".pfm", fmWrite)
    oImg.writePfmImage(outPfm)
    echo "HDR image written to " & outputFileName & ".pfm"
    outPfm.close()

    oImg.normalizeImage(factor)
    oImg.clampImage()

    oImg.writeLdrImage(outputFileName & ".png", gamma)

    echo ("File " & outputFileName & " has been written to disk")

# --------------PFM2FORMAT--------------

proc pfm2format(inPfmFileName: string, factor = 0.2, gamma = 1.0,
        outputFileName: string) =

    let inPfm = newFileStream(inPfmFileName, fmRead)
    var img = readPfmImage(inPfm)
    inPfm.close()

    echo ("File " & inPfmFileName & " has been read from disk")

    img.normalizeImage(factor)
    img.clampImage()

    img.writeLdrImage(outputFileName, gamma)

    echo ("File " & outputFileName & " has been written to disk")

# --------------DEMO (Deprecated)--------------

proc demo(angleDeg = 0.0, orthogonal = false, width = 640, height = 480,
        fileName = "demo", format = "png", algorithm = "pathtracing",
        raysNum = 10, maxDepth = 3, initState = 42, initSeq = 54,
                samplePerPixel = 1.0,
        luminosity: float = 0.0) =

    let samplesPerSide = sqrt(samplePerPixel).int
    if pow(samplesPerSide.float, 2.float) != samplePerPixel:
        quit("Error, the number of samples per pixel ({samplePerPixel}) must be a perfect square")

    var image = newHDRImage(width, height)
    echo("Generating a ", width, "x", height,
            " image, with the camera tilted by ", angleDeg, "°")

    var world = newWorld()
    var cameraTr : Transformation

    if (algorithm != "on/off"):
        let skyMaterial = newMaterial(
            brdf = newDiffuseBRDF(pigment = newUniformPigment(newColor(0.7, 0.7, 0.7))),
            emittedRadiance = newUniformPigment(newColor(0.7, 0.7, 0.7))
            )

        let groundMaterial = newMaterial(
            brdf = newDiffuseBRDF(
                pigment = newCheckeredPigment(
                    color1 = newColor(0.3, 0.5, 0.1),
                    color2 = newColor(0.1, 0.2, 0.5)))
            )

        let inPfm = newFileStream("img/pfm2formatExamples/sample.pfm", fmRead)
        var img = readPfmImage(inPfm)
        inPfm.close()


        let bigSphereMaterial = newMaterial(
            brdf = newDiffuseBRDF(
                pigment = newImagePigment(img))
            )
        

        let littleSphereMaterial = newMaterial(brdf = newDiffuseBRDF(
                pigment = newUniformPigment(newColor(0.7, 0.1, 0.3)))
        )

        #[
        let mirrorMaterial = newMaterial(
            brdf = newSpecularBRDF(
                pigment = newUniformPigment(newColor(0.6, 0.2, 0.3))
            )
        )
        ]#

        world.addShape(
            newSphere(material = skyMaterial,
            transformation = scaling(newVec(200, 200, 200)) * translation(newVec(0,
                    0, 0.4)))
        )

        world.addShape(newPlane(material = groundMaterial))

        world.addShape(
            newSphere(
                material = bigSphereMaterial,
                transformation = translation(newVec(0, 0, 1))
            )
        )

        world.addShape(
            newSphere(
                material = littleSphereMaterial,
                transformation = translation(newVec(0,0, 2.5))*scaling(newVec(0.3,0.3,0.3))
                # Pay attention to the order of the transformations!
            )
        )
 
        world.addLight(newPointLight(position = newPoint(-30, 30, 30),
                color = white))
        

        # Define transformation on camera
        cameraTr = rotationZ(angleDeg) * translation(newVec(-2.0, 0.0, 1.0))

    else:
        # Add spheres as vertices of a 0.5 side cube
        for x in [-0.5, 0.5]:
            for y in [-0.5, 0.5]:
                for z in [-0.5, 0.5]:
                    world.addShape(
                        newSphere(
                            transformation = translation(newVec(x, y, z))*scaling(
                                    newVec(0.1, 0.1, 0.1))
                        )
                    )

        # Place two other spheres in the cube, in order to check whether
        # there are issues with the orientation of the image

        # First sphere at bottom
        world.addShape(
            newSphere(
                transformation = translation(newVec(0.0, 0.0, -0.5)) *
                    scaling(newVec(0.1, 0.1, 0.1))
            )
        )

        # Second sphere on the left face
        world.addShape(
            newSphere(
                transformation = translation(newVec(0.0, 0.5, 0.0)) *
                    scaling(newVec(0.1, 0.1, 0.1))
            )
        )

        cameraTr = rotationZ(angleDeg) * translation(newVec(-1.0, 0.0, 0.0))

    var camera: Camera

    if orthogonal:
        camera = newOrthogonalCamera(aspectRatio = width / height,
                transformation = cameraTr)
    else:
        camera = newPerspectiveCamera(aspectRatio = width / height,
                transformation = cameraTr)

    # Run the tracer

    var tracer = newImageTracer(image, camera, samplesPerSide = samplesPerSide)

    var renderer: Renderer

    case algorithm:
        of "on/off":
            echo("Using on/off renderer")
            renderer = newOnOffRenderer(world, black, white)
        of "flat":
            echo("Using flat renderer")
            renderer = newFlatRenderer(world)
        of "pathtracing":
            echo("Using pathtracing")
            var pcg = newPCG(initState = initState.uint64,
                    initSeq = initSeq.uint64)
            renderer = newPathTracer(
                world = world,
                pcg = pcg,
                raysNum = raysNum,
                maxDepth = maxDepth
            )
        of "pointlight":
            echo("Using point-light tracer")
            renderer = newPointLightRenderer(world = world,
                    backgroundColor = black)
        else:
            quit("Unknown renderer")

    let time = getMonoTime()
    tracer.fireAllRays(ray => call(renderer, ray))
    echo "Time taken: ", getMonoTime() - time

    # Save the HDR image
    let outPfm = newFileStream(fileName & ".pfm", fmWrite)
    tracer.image.writePfmImage(outPfm)
    echo "HDR demo image written to " & fileName & ".pfm"
    outPfm.close()

    # Apply tone-mapping to the image
    if luminosity == 0.0:
        tracer.image.normalizeImage(factor = 1.0) #use average luminosity
    else:
        tracer.image.normalizeImage(factor = 1.0, luminosity)

    tracer.image.clampImage()

    # Save the LDR image
    tracer.image.writeLdrImage(fileName & "." & format)
    echo "PNG demo image written to " & fileName & ".png"

# --------------MAIN--------------

when isMainModule:

    const traynim = """
 
                     _______*             _   _ _           
                    |__   __ *           | \ | (_)          
                       | |_ *_ __ _ _   _|  \| |_ _ __ ___  
                       | | '__/ _` | | | | . ` | | '_ ` _ \ 
                       | | | | (_| | |_| | |\  | | | | | | |
                       |_|_|  \__,_|\__, |_| \_|_|_| |_| |_|
                                     __/ |                  
                                    |___/                   
                    """

    echo traynim

    const rHelp = { "inSceneName" : "name of the file containing the image",
                    "declareFloat": "Declare a variable. The syntax is «--d=VAR:VALUE». Example: --d=\"clock:150 ang:90\"",
                    "width": "Width of the image",
                    "height": "Height of the image",
                    "fileName": "Path to output file without format",
                    "format": "PNG, PPM, BMP or QOI formats",
                    "algorithm": "pathtracing, pointlight, flat or on/off renderer",
                    "luminosity": "luminosity for LDR image conversion, lower number is lighter, default is averageLuminosity",
                    "samplesPerPixel": "Number of samples per pixel (this must be a perfect square, e.g., 16)",
                    "raysNum" : "Number of rays departing from each surface point (only applicable with --algorithm=pathtracing)",
                    "maxDepth": "Maximum allowed ray depth (only applicable with --algorithm=pathtracing)",
                    "initState" : "Initial seed for the random number generator (positive number)",
                    "initSeq" : "Identifier of the sequence produced by the random number generator (positive number)"}.toTable

    const rShort = { "inSceneName" : 'i',
                    "height": 'e',
                    "fileName" : 'o',
                    "format": 'f',
                    "initState" : 't',
                    "initSeq" : 'q'}.toTable

    dispatchMulti(
        [renderer, help = rHelp, short=rShort],
        [stack],
        [pfm2format,
            help = {"outputFileName": " Path to output file (PNG, PPM, BMP or QOI formats)",
                    "factor": "Multiplicative factor",
                    "gamma": "Exponent for gamma-correction",
                    "inPfmFileName": "Path to input file (PFM)"}
        ],
        [demo,
            help = {"angleDeg": "Angle rotation of the camera (Degrees)",
                    "orthogonal": "Perspective or orthogonal camera (fefault is perspective)",
                    "width": "Width of the image",
                    "height": "Height of the image",
                    "fileName": "Path to output file without format",
                    "format": "PNG, PPM, BMP or QOI formats",
                    "samplePerPixel": "Number of samples per pixel (must be a perfect square, e.g., 16)",
                    "algorithm": "options on/off, flat, pathtracing renderer",
                    "luminosity": "luminosity for LDR image conversion, lower number is lighter, default is averageLuminosity"}
        ])
